src/julia_gc.c

```diff
@@ -0,0 +1,755 @@
+        InfoBags[new_type].sizeLive += size;
+        InfoBags[old_type].sizeAll -= size;
+        InfoBags[new_type].sizeAll += size;
+    }
+#endif
+
+    header->type = new_type;
+}
+
+static inline Bag AllocateMasterPointer(void)
+{
+    // HOOK: Allocate memory for the master pointer.
+    // Master pointers require one word of memory.
+    void * result =
+        (void *)jl_gc_alloc(JuliaTLS, sizeof(void *), datatype_mptr);
+    memset(result, 0, sizeof(void *));
+    return result;
+}
+
+Bag NewBag(UInt type, UInt size)
+{
+    Bag  bag; /* identifier of the new bag       */
+    UInt alloc_size;
+
+    alloc_size = sizeof(BagHeader) + size;
+
+    SizeAllBags += size;
+
+    /* If the size of an object is zero (such as an empty permutation),
+     * and the header size is a multiple of twice the word size of the
+     * architecture, then the master pointer will actually point past
+     * the allocated area. Because this would result in the object
+     * being freed prematurely, we will allocate at least one extra
+     * byte so that the master pointer actually points to within an
+     * allocated memory area.
+     */
+    if (size == 0)
+        alloc_size++;
+
+    BagHeader * header = AllocateBagMemory(type, alloc_size);
+
+    header->type = type;
+    header->flags = 0;
+    header->size = size;
+
+    bag = AllocateMasterPointer();
+
+    /* set the masterpointer                                               */
+    SET_PTR_BAG(bag, DATA(header));
+    /* return the identifier of the new bag                                */
+    return bag;
+}
+
+UInt ResizeBag(Bag bag, UInt new_size)
+{
+    UInt  type; /* type of the bag                 */
+    UInt  flags;
+    UInt  old_size; /* old size of the bag             */
+    Bag * src;      /* source in copying               */
+    UInt  alloc_size;
+
+    BagHeader * header = BAG_HEADER(bag);
+
+    /* get type and old size of the bag                                    */
+    type = header->type;
+    flags = header->flags;
+    old_size = header->size;
+
+#ifdef COUNT_BAGS
+    /* update the statistics                                               */
+    InfoBags[type].sizeLive += new_size - old_size;
+    InfoBags[type].sizeAll += new_size - old_size;
+#endif
+    SizeAllBags += new_size - old_size;
+
+    if (new_size <= old_size) {
+        /* change the size word                                            */
+        header->size = new_size;
+    }
+
+    /* if the bag is enlarged                                              */
+    else {
+        alloc_size = sizeof(BagHeader) + new_size;
+        if (new_size == 0)
+            alloc_size++;
+        header = AllocateBagMemory(type, alloc_size);
+
+        header->type = type;
+        header->flags = flags;
+        header->size = new_size;
+
+        // copy data and update the masterpointer
+        src = PTR_BAG(bag);
+        memcpy(DATA(header), src, old_size < new_size ? old_size : new_size);
```

Comment:
even simpler would be to copy *everything, including the `BagHeader`, and then only update `header->size`